### PR TITLE
Report Orders and Distributors should display variant

### LINKED ISSUE #12878: Report Orders and Distributors should display variant
## Description
This report should display the product variant. Instead the "variant" column displays the unit. So this report can not be used by shops who use variants.

## Expected Behavior
Report should display the product variant.

## Actual Behaviour
Only Product, unit... are included in report.

## Steps to Reproduce
Ask a report where some products have various variants.

## Workaround
Unkown

## Severity
s3 ?
Together with https://github.com/openfoodfoundation/openfoodnetwork/issues/12877 hubs have some difficulty using various delivery methods in a same order cycle.

### PR DESCRIPTION
**Pull request by chahmedejaz**
#### What? Why?

- Closes #12878
- This PR adds the report variant name under the "Variant" column to the Order and Distributor report to mimic what's been displayed on the orders page for the variant.
- Moreover, after changing the item name to "Product", existing selected column name with `item_name` selected, won't be selected anymore. So to keep the selected column names the same in this case, a data migration has been added which changes the selected item_name to the product in all the existing selected columns that already had selected item_name

#### What should we test?

- As described or discussed in the issue
- Moreover, please make sure that if the item name column was selected before deployment, then after deployment product column name would have to be selected or vise versa

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [ ] Feature toggled

### CODE REVIEW

File: db/migrate/20241011071014_update_item_name_to_product_in_od_report.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    options = ReportRenderingOptions.where(report_type: 'orders_and_distributors', report_subtype: nil)
+
+    options.find_each do |option|
+      begin
+        fields_to_show = option.options[:fields_to_show]
+        next if fields_to_show&.exclude?('item_name')
+
+        fields_to_show.delete('item_name')
+        fields_to_show << 'product'
+        option.update(options: option.options)
```
Comment (dacook) [inline db/migrate/20241011071014_update_item_name_to_product_in_od_report.rb:14]: I guess this works if `fields_to_show` is a pointer to the value inside the options hash. 
If that's the case, I think you should be able to simply call:
```suggestion
        option.save
```

But either way, I'm not sure as it's an uncommon way to do things, so I would like to see a test to check that it works.

**Review by dacook — CHANGES_REQUESTED** [review]
Great work, and good thinking to migrate the options data to make it seamless for users.

I have a concern about the migration, do you think it's worth adding a test? Or are you confident you've already manually checked all possibilities?

**Comment by chahmedejaz** [comment]
> I have a concern about the migration, do you think it's worth adding a test? Or are you confident you've already manually checked all possibilities?

@dacook - I'd manually tested all the possible use cases for this migration, I thought it would run only once so wasn't sure if it was worth a spec that would add up to the pipeline each time the workflow runs.
However, yes, I may have missed something in my manual testing, so I've added the specs for the use cases I could think of. Please review and let me know what you think about it. Thanks.

**Review by dacook — APPROVED** [review]
Wonderful, thank you.

Yes it's hard to know exactly what to spec or not. It's still production code that will modify data on every server, so I thought it worth being certain about exactly what it will and won't do.

I agree we don't need the test running every time, and maybe it would be possible to configure a runner to only test migrations when they are added. But I don't know if there's an easy way to do that. For now, I think we just occasionally archive the migrations and specs.

**Review by rioug — APPROVED** [review]
Looks good :+1:

**Comment by filipefurtad0** [comment]
Hey @chahmedejaz ,

Unfortunately, staging failed on both staging servers:

- staging-AU
https://github.com/openfoodfoundation/openfoodnetwork/actions/runs/11827139747/job/32954654935

and

- staging-FR
https://github.com/openfoodfoundation/openfoodnetwork/actions/runs/11827025386/job/32954294829

Moving back to In Progress for now.

**Comment by chahmedejaz** [comment]
Thanks for raising this Filipe. I hope it will get fixed now.

@dacook - Here's the fix:
https://github.com/openfoodfoundation/openfoodnetwork/pull/12930/commits/39fa8e0acecb9f7a955b45735bd904988927d8dd

The migration class name was a bit different than the file name...Seems like I forgot to run the migration again after changing the name or something. 😞

**Review by dacook — APPROVED** [review]
Cool, looks like there has been a newer migration since this one was created, but that should be ok.

**Comment by chahmedejaz** [comment]
Thanks @dacook for the quick response, are we good to move it to testing or do we need a second review? Thanks.

**Comment by filipefurtad0** [comment]
Hey @dacook and @chahmedejaz ,

I've tried several options to stage this PR, but even after resetting the DB with the suggested command (`~/apps/openfoodnetwork/current/script/ci/restore-staging-baseline.sh`) it keeps failing:

https://github.com/openfoodfoundation/openfoodnetwork/actions/runs/11884093947/job/33111772141

Any ideas?

Thanks again.

**Review by dacook — CHANGES_REQUESTED** [review]
Ah, sorry I missed that there's a problem with the migration. This should have been picked up  during dev or CI testing, I'm not sure why it was missed.

Ahmed can you please look into this? The Good Migrations gem is linting our migrations to ensure they don't refer to application code. Migrations need to be completely independent, to ensure they always behave the same, no matter when they are executed.

In this case, I think you can simply define a basic ActiveModel class for table ReportRenderingOptions. See previous migrations for an example.

**Comment by chahmedejaz** [comment]
Sure @dacook, it's done and pushed. Please review. I remember validating this migration before pushing to validate the case that the migration class is picked up successfully now, that's strange it didn't give this error then 😅 
Strangely, it didn't raise any error during the CI build as well.
Anyhow, hope it works now. Thanks

**Review by dacook — APPROVED** [review]
Thanks

**Comment by dacook** [comment]
FYI @filipefurtad0 I've started deployment to AU staging.
There's one CI failure so far but I think it's a flaky one and can be ignored.

**Comment by chahmedejaz** [comment]
CI specs got broken for this migration.
This is weird. This spec was working just before this change. Let me take a quick look

**Comment by chahmedejaz** [comment]
The spec fails with the Basic ActiveRecord table. Strange again... Looking

**Comment by chahmedejaz** [comment]
@dacook @filipefurtad0 - It's all green now. Sorry for the back and forth.

**Review by dacook — APPROVED** [review]

**Comment by filipefurtad0** [comment]
Hey again @chahmedejaz @dacook ,

I've placed an order with different variants and checked the report, before this PR:
image

and after this PR:
image

It now displays the variant names, so I believe it makes the report usable again.

Merging!!
Thanks for your effort :muscle:

**Pull request merged**